Automattic/wp-calypso
# Replace morgan with bunyan to produce logs in JSON format

### Background

In order to improve log coverage on the server, we need to send the logs to Elasticsearch. But ES only consumes logs in JSON format, so we need to add more structure to them.

### Changes

* Replace `morgan` with `bunyan` to produce access logs in JSON format. I tried to log the same info than the previous formats (`dev` and `combined` for `development` and `production`, respectively) by leveraging `morgan` parser functions.


### Testing instructions

* Checkout this branch and run `yarn start`. You should see logs in structured (*) format:
    ```
    12:46:43.279Z  INFO calypso:  (method=GET, status=304, duration=0.363)
        url: /calypso/evergreen/async-load-blocks-support-article-dialog.js
    12:46:43.281Z  INFO calypso:  (method=GET, status=304, duration=0.437)
        url: /calypso/evergreen/async-load-blocks-legal-updates-banner.css
    12:46:43.293Z  INFO calypso:  (method=GET, status=304, duration=0.36)
    ```
    (*) when running `yarn start` I'm using `bunyan` CLI tool to prettify the logs, that's why they they don't look like JSON.

* Stop the server and run it with `NODE_ENV=production yarn start`, you should see more fields in the logs:
    ```
    12:56:04.229Z  INFO calypso:  (method=GET, status=404, length=4150, url=/__webpack_hmr, duration=740.129, remoteAddr=::1, httpVersion=1.1, referrer=http://calypso.localhost:3000/)
        userAgent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.83 Safari/537.36
    ```

* Stop the server and run it again with `node build/server.js`. Now you should see a pure JSON log.
    ```
    {"name":"calypso","hostname":"sergio.lan","pid":59465,"level":30,"method":"GET","status":"404","length":"4150","url":"/calypso/evergreen/vendors~entry-login~entry-main.js","duration":12.282,"remoteAddr":"::1","httpVersion":"1.1","referrer":"http://calypso.localhost:3000/","userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.83 Safari/537.36","msg":"","time":"2020-09-10T12:56:53.743Z","v":0}
    ```

### Not covered

This PR does not cover transforming other logs into JSON format (eg: usages of console.log still produce raw text lines). That will be addressed in a future PR to keep things simpler.
